lib/ansible/modules/cloud/vmware/vmware_guest.py

```diff
@@ -1479,19 +1488,15 @@ def main():
             result = pyv.reconfigure_vm()
         elif module.params['state'] in ['poweredon', 'poweredoff', 'restarted', 'suspended', 'shutdownguest', 'rebootguest']:
             # set powerstate
-            tmp_result = pyv.set_powerstate(vm, module.params['state'], module.params['force'])
-            if tmp_result['changed']:
-                result["changed"] = True
-            if not tmp_result["failed"]:
-                result["failed"] = False
+            result = pyv.set_powerstate(vm, module.params['state'], module.params['force'])
```

Comment:
This will replace everything in the result dict. This is not correct.

I see now that the logic is wrong, and I am not sure when or why this is like this.
In all states (except absent) we should perform a reconfigure. That is why in the case of a specific power state, it may not replace the result dict.

Reply:
But, the `result` was defined as `result = {'failed': False, 'changed': False}` and it will not be changed before here. Actually `set_powerstate()` will also return a data which has the same structure as `result`.

It seems that performing reconfigure isn't necessary to change the power state, what we should do is just calling the methods of `vm` like `vm.PowerOff()`.

So, I think it will not cause some problems or confusion in here.

PS: If I lost something, please tell me :-)

Reply:
So most people will declare the VMs that should exist, with the power-state it should be in.

If a user then reconfigures the declaration for that VM, and e.g. changes something to the VM, the powerstate will still be stated and he will expect the VM to be reconfigured, and powered on if it wasn't powered on.

So yes, the module should be able to reconfigure and change power state in a single call.

And because of the above, we should not simply use the result of the set_powerstate function, but we have to merge some of the return values, as was originally done.

I think someone changed **vmware_guest**, not understanding people may want to make a change to a VM and setting power state.

Reply:
So, that means it should be changed like the following?
```diff
@@ -1482,8 +1493,12 @@ def main():
             tmp_result = pyv.set_powerstate(vm, module.params['state'], module.params['force'])
             if tmp_result['changed']:
                 result["changed"] = True
-            if not tmp_result["failed"]:
-                result["failed"] = False
+            if tmp_result['failed']:
+                result['failed'] = True
+            if tmp_result['msg']:
+                result['msg'] = tmp_result['msg']
+            if tmp_result['instance']:
+                result['instance'] = tmp_result['instance']
```

**Note**:
- The VM instance will be returned when `tmp_result['failed']` isn't `True`.
- `tmp_result['msg']` will be merged to the `result`. The original way just ignored it.
- `result['failed']` should be set to `True` when `tmp_result['failed']` is `True`. The original way maybe cause `result['failed']` always is `False` even if setting power state was failed.